File: app/src/main/java/com/commit451/gitlab/adapter/BreadcrumbAdapter.java
package com.commit451.gitlab.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.commit451.gitlab.R;
import com.commit451.gitlab.model.Breadcrumb;
import com.commit451.gitlab.viewHolders.BreadcrumbViewHolder;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Shows the current file path
 * Created by Jawnnypoo on 11/22/2015.
 */
public class BreadcrumbAdapter extends RecyclerView.Adapter<BreadcrumbViewHolder> {
    private List<Breadcrumb> mValues;

    public BreadcrumbAdapter() {
        mValues = new ArrayList<>();
        notifyDataSetChanged();
    }

    private final View.OnClickListener onProjectClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            int position = (int) v.getTag(R.id.list_position);
            Breadcrumb breadcrumb = getValueAt(position);
            if (breadcrumb != null && breadcrumb.getListener() != null) {
                breadcrumb.getListener().onClick();
            }
        }
    };

    @Override
    public BreadcrumbViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        BreadcrumbViewHolder holder = BreadcrumbViewHolder.newInstance(parent);
        holder.itemView.setOnClickListener(onProjectClickListener);
        return holder;
    }

    @Override
    public void onBindViewHolder(final BreadcrumbViewHolder holder, int position) {
        String title = "";
        boolean showArrow = position != mValues.size() - 1;

        Breadcrumb breadcrumb = getValueAt(position);
        if (breadcrumb != null) {
            title = breadcrumb.getTitle();
        }

        holder.bind(title, showArrow);
        holder.itemView.setTag(R.id.list_position, position);
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    public void setData(Collection<Breadcrumb> breadcrumbs) {
        mValues.clear();
        if (breadcrumbs != null) {
            mValues.addAll(breadcrumbs);
            notifyItemRangeInserted(0, breadcrumbs.size());
        }
        notifyDataSetChanged();
    }

    public Breadcrumb getValueAt(int position) {
        if (position < 0 || position >= mValues.size()) {
            return null;
        }

        return mValues.get(position);
    }
}

File: app/src/main/java/com/commit451/gitlab/adapter/BreadcrumbAdapter.kt
package com.commit451.gitlab.adapter

import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import com.commit451.gitlab.R
import com.commit451.gitlab.viewHolder.BreadcrumbViewHolder
import java.util.*

/**
 * Shows the current file path
 */
class BreadcrumbAdapter : androidx.recyclerview.widget.RecyclerView.Adapter<BreadcrumbViewHolder>() {
    private val values: MutableList<Breadcrumb>

    init {
        values = ArrayList<Breadcrumb>()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BreadcrumbViewHolder {
        val holder = BreadcrumbViewHolder.inflate(parent)
        holder.itemView.setOnClickListener { v ->
            val position = v.getTag(R.id.list_position) as Int
            val breadcrumb = getValueAt(position)
            if (breadcrumb != null) {
                breadcrumb.listener.onClick()
            }
        }
        return holder
    }

    override fun onBindViewHolder(holder: BreadcrumbViewHolder, position: Int) {
        var title = ""
        val showArrow = position != values.size - 1

        val breadcrumb = getValueAt(position)
        if (breadcrumb != null) {
            title = breadcrumb.title
        }

        holder.bind(title, showArrow)
        holder.itemView.setTag(R.id.list_position, position)
    }

    override fun getItemCount(): Int {
        return values.size
    }

    fun setData(breadcrumbs: Collection<Breadcrumb>?) {
        values.clear()
        if (breadcrumbs != null) {
            values.addAll(breadcrumbs)
            notifyItemRangeInserted(0, breadcrumbs.size)
        }
        notifyDataSetChanged()
    }

    fun getValueAt(position: Int): Breadcrumb? {
        if (position < 0 || position >= values.size) {
            return null
        }

        return values[position]
    }

    class Breadcrumb(val title: String, val listener: Listener)

    interface Listener {
        fun onClick()
    }
}
